### PR TITLE
Add ETC2 textures support.

### PR DESCRIPTION
**Pull request by sssooonnnggg**
#2181 

Add ETC2 textures support for mobile devices and mobile browsers (WebGL).

### CODE REVIEW

**Comment by eugeneko** [comment]
I'm ok with this PR.
But commits will be squashed before merge.

**Comment by sssooonnnggg** [comment]
OK, commits are squashed.

File: Source/Urho3D/Resource/DecompressETC2.h [hunk 1]
```diff
@@ -0,0 +1,14 @@
+
```
Comment (weitjong) [inline Source/Urho3D/Resource/DecompressETC2.h]: The header and implementation files are missing the Urho3D copyright notice. Also, just for consistency sake, I think all the newly added functions should be declared using `Urho3D` namespace.

File: Source/Urho3D/Resource/Image.cpp [hunk 2]
```diff
@@ -704,6 +725,16 @@ bool Image::BeginLoad(Deserializer& source)
...
+        case 22:
+            compressedFormat_ = CF_ETC2_RGB;
+            components_ = 3;
+            break;
+
+        case 23:
+            compressedFormat_ = CF_ETC2_RGBA;
+            components_ = 4;
+            break;
+
```
Comment (weitjong) [inline Source/Urho3D/Resource/Image.cpp]: I actually wanted to ask why there is duplication here but forgot to ask it in the earlier review. Care to enlighten us? Perhaps you might as well explain the reason as code comment, if they are intentional.

**Comment by 1vanK** [comment]
Why did you create a new file DecompressETC2.h? Why not just add code to Decompress.h

**Comment by 1vanK** [comment]
We already have ETC supputs. I don’t know about this format, but are you sure that they have nothing in common? It looks like you're re-writing functions to parse all data blocks (Sorry if I'm wrong, I glanced at the code briefly)

**Comment by sssooonnnggg** [comment]
@1vanK The old decompressETC code can't support the ETC2(diffrerent from ETC1) format(with or without alpha channel),and is also copied from another library.

Decompressetc2.cpp code is from libctx, which is not a part of libktx's interface but some kind of its implementation. I think it is not necessary to add the full library to our code base.Maybe need more license file?

File: Source/Urho3D/Resource/Image.cpp (hunk 2 above)
Comment (sssooonnnggg) [reply Source/Urho3D/Resource/Image.cpp]: @weitjong  Both .ktx files and .pvr files can support etc2 image format, their difference is just the header.

**Comment by weitjong** [comment]
IANAL but since 1vanK has already brought the thing up then it needs to be looked into now. The part that states "it is in unmodified form" is problematic. I think moving that source file in its original name unmodified to the 3rd-party libs is the right way to go. We can easily add the original license file there like all our 3rd-party deps.

File: Source/Urho3D/Resource/Image.cpp (hunk 2 above)
Comment (weitjong) [reply Source/Urho3D/Resource/Image.cpp]: I see. If you could add a short comment line to say which is which that would be great.

**Comment by sssooonnnggg** [comment]
OK, I will move that file to.rhe 3rd-party libs.

**Comment by rokups** [comment]
Hey @sssooonnnggg this is a great PR. Absolutely love it that you added decompression code which i did not bother with.  I noticed that gist you linked also has ETC1 decompression. If you are including that maybe it would be worthwhile to also make ETC1 decompression use your included lib and delete current ETC1 decompression code?

**Comment by sssooonnnggg** [comment]
@rokups Good suggestion, I will test new decompression code on ETC1 format and then delete current ETC1 decompression code. :)

**Comment by sssooonnnggg** [comment]
I have add ETCPACK as a 3rd-party lib and use it to decompress both ETC1/ETC2 texture, it works properly now.

File: Source/Urho3D/Resource/Decompress.cpp [hunk 3]
```diff
@@ -269,187 +278,7 @@ void DecompressImageDXT(unsigned char* rgba, const void* blocks, int width, int
...
-                    }
-                }
-            }
-
-            // advance
-            sourceBlock += bytesPerBlock;
-        }
-    }
-}
+// PVRTC decompression based on the Oolong Engine, modified for Urho3D
```
Comment (eugeneko) [inline Source/Urho3D/Resource/Decompress.cpp]: Is this comment relevant?

Comment (sssooonnnggg) [reply Source/Urho3D/Resource/Decompress.cpp]: It just changed comment
`// ETC and PVRTC decompression based on the Oolong Engine, modified for Urho3D`
To
`// PVRTC decompression based on the Oolong Engine, modified for Urho3D`

File: Source/CMakeLists.txt [hunk 4]
```diff
@@ -68,7 +68,7 @@ if (URHO3D_LIB_TYPE STREQUAL SHARED AND NOT WIN32 AND NOT ANDROID AND NOT IOS AN
 endif ()
 
 # Add targets
-foreach (TARGET FreeType LZ4 PugiXml rapidjson SDL StanHull STB)
+foreach (TARGET FreeType LZ4 PugiXml rapidjson SDL StanHull STB ETCPACK)
```
Comment (weitjong) [inline Source/CMakeLists.txt]: Keep the list sorted, please.

File: Source/ThirdParty/ETCPACK/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+# Define target name
+set (TARGET_NAME ETCPACK)
+
+# Define source files
+set (SOURCE_FILES "source/etcdec.cxx")
```
Comment (weitjong) [inline Source/ThirdParty/ETCPACK/CMakeLists.txt]: For consistency with the rest of the CMake build script, remove the unnecessary double quotes.

**Comment by weitjong** [comment]
Also, need to append the EXTPACK license to the https://github.com/urho3d/Urho3D/blob/master/Source/ThirdParty/LICENSES

**Comment by weitjong** [comment]
Thanks,

**Pull request merged**